### PR TITLE
Fix message races

### LINKED ISSUE #2129: TestService_DownloadStateRunning fails intermittently
Jenkins reports that `TestService_DownloadStateRunning` fails from time to time.

And there seems also to be a race-condition while doing `TestClose`.

### PR DESCRIPTION
**Pull request by ineiti**
Divers races have been fixed:
- between closing down byzcoin/skipchain and calling propagateForwardLinks
- between resetting byzcoin and calling propagateForwardLinks
- added some `local.WaitDone(time.Second)` to the end of some tests to make sure all messages are passed before closing the services

Plus a proposed change to the ill-named skipchain::Service.WaitBlock

Closes #2129

---

🙅‍ Friendly checklist:

- [x] 0. Code comments are added (or updated) when/where needed and explain the WHY of the code.
- [x] 1. Design choices, user documentation and any additional doc are added (or updated) in READMEs.
- [x] 2. Any new behaviour is tested and small units of code that can be are unit tested.
- [x] 3. Code comments are added on tests to explain what they do.
- [x] 4. Errors are systematically wrapped with a meaningful message using `xerrors.Errorf` and the `%v` verb.
- [x] 5. Hard limit of 80 chars is always respected.
- [x] 6. Changes are backward compatible.
- [x] 7. Indentation level does not exceed 5, although 4 is already suspicious.
- [x] 8. Functions, files, and packages are kept to a manageable size and decomposed into smaller units if needed.
- [x] 9. There are no magic values.

### CODE REVIEW

File: skipchain/skipchain.go [hunk 1]
```diff
@@ -955,9 +955,24 @@ func (s *Service) ListFollow(list *ListFollow) (*ListFollowReply, error) {
...
+// ChainIsProcessing returns true if the given chain has a block in the
+// blockBuffer.
+func (s *Service) ChainIsProcessing(sid SkipBlockID) bool {
+	return s.blockBuffer.has(sid)
+}
+
 // WaitBlock returns a block by its ID instantly if already stored in the DB
 // or check if the block is inside the buffer. If the block is known, false will
 // be returned because a catch up is not necessary, true otherwise.
+// DEPRECATED
```
Comment (nkcr) [inline skipchain/skipchain.go]: > To signal that an identifier should not be used, add a paragraph to its doc comment that begins with Deprecated: followed by some information about the deprecation, and a recommendation on what to use instead, if applicable.

I would add some explanations about the deprecation

File: byzcoin/service.go [hunk 2]
```diff
@@ -2569,7 +2579,9 @@ func (s *Service) cleanupGoroutines() {
 		delete(s.pollChan, k)
 	}
 	s.pollChanMut.Unlock()
+	log.Lvl1(s.ServerIdentity(), "closing go-routines 2")
 	s.pollChanWG.Wait()
+	log.Lvl1(s.ServerIdentity(), "closing go-routines 3")
```
Comment (tharvik) [inline byzcoin/service.go]: Debug?

File: byzcoin/service_test.go [hunk 3]
```diff
@@ -2709,8 +2710,8 @@ func (s *ser) testDarcEvolution(t *testing.T, d2 darc.Darc, fail bool) (pr *Proo
 }
 
 func (s *ser) deleteDBs(t *testing.T, index int) {
-	log.Lvl1("Deleting DB of node", index)
 	bc := s.services[index]
+	log.Lvl1("Deleting DB of node", index, bc.ServerIdentity())
```
Comment (tharvik) [inline byzcoin/service_test.go]:
```suggestion
	log.Lvlf1("%v: deleting DB", bc.ServerIdentity())
```

File: byzcoin/tx_pipeline.go [hunk 4]
```diff
@@ -110,7 +110,11 @@ func (s *defaultTxProcessor) CollectTx() (*collectTxResult, error) {
 		return nil, xerrors.Errorf("reading config: %v", err)
 	}
 
-	_, isNotProcessingBlock := s.skService().WaitBlock(s.scID, nil)
+	if s.skService().ChainIsProcessing(s.scID) {
+		// When a block is processed, wait for the block to be done
+		time.Sleep(bcConfig.BlockInterval / 2)
+		return &collectTxResult{Txs: nil, CommonVersion: CurrentVersion - 1}, nil
```
Comment (tharvik) [inline byzcoin/tx_pipeline.go]: That looks weird, you wait for some time then return the same data you'd have before sleeping.

Comment (ineiti) [reply byzcoin/tx_pipeline.go]: Looking at the call chain, I think I can remove that one. Sending it to Travis for confirmation ;)

**Pull request merged**